cylc/flow/resources.py

```diff
@@ -14,99 +14,180 @@
+    sessions.
+    """
+    keys = []
+    with open((TUTORIAL_DIR / 'api-keys'), 'r') as api_keys:
+        for api_key in api_keys:
+            keys.append(api_key)
+    shuffle(keys)
+    return keys[0]
+
+
+def set_api_key(tgt):
+    """Replace a placeholder with a real API key.
+
+    Replaces the placeholder DATAPOINT_API_KEY with a value chosen at random
+    from the file api-keys chosen.
     """
-    if resources is None:
-        resources = resource_names
-    for resource in resources:
-        if (
-            resource not in resource_names
-            and resource != 'etc/tutorial'
-        ):
-            raise ValueError(f"Invalid resource name {resource}")
-        path = Path(target_dir) / Path(resource).name
-        LOG.info(f"Extracting {resource} to {path}")
-        pdir = path.parent
-        if not pdir.exists():
-            pdir.mkdir(parents=True)
-        # In spite of the name, this returns a byte array, not a string:
-        res = pr.resource_string('cylc.flow', resource)
-        with open(path, 'wb') as h:
-            h.write(res)
-
-
-def extract_tutorials():
-    """Extract tutorial workflows to Cylc Source Directory."""
-    src = Path(cylc.flow.__file__).parent / 'etc/tutorial'
-    dest = glbl_cfg().get(['install', 'source dirs'])[0]
-    dest = Path(dest).expanduser() / 'cylc-tutorials'
-    if dest.exists():
-        # If destination exists timestamp it and replace it.
-        tstamp = get_current_time_string(use_basic_format=True)
-        backup = dest / f'{tstamp}.cylc-tutorials'
-        shutil.copytree(dest, backup)
-        shutil.rmtree(dest)
-        LOG.warning(
-            'Replacing an existing cylc-tutorials folder which will'
-            f' be copied to {backup}')
-    shutil.copytree(src, dest)
-    LOG.info(f'Cylc tutorials extracted to {str(dest)}')
+    # get the api key
+    api_key = get_api_key()
+    # go through all the top level files
+    for path in tgt.glob('*'):
+        if not path.is_dir():
+            # write the file out one line at a time to a temp file
+            tmp_path = path.parent / (path.name + '.tmp')
+            with open(path, 'rb') as _src, open(tmp_path, 'wb+') as _tmp:
+                # NOTE: open the file in bytes mode for safety
+                # (prevents decode errors surfacing here)
+                for line in _src:
+                    _tmp.write(
+                        # perform the replacement line by line
+                        # (some things are easier with sed!)
+                        line.replace(
+                            b'DATAPOINT_API_KEY',
+                            api_key.encode(),
+                        )
+                    )
+
+            # then move the tmpfile over the original
+            # NOTE: shutil interfaces don't fully support Path objects at all
+            # python versions
+            path.unlink()
+            shutil.move(str(tmp_path), str(path))
```

Comment:
Why not simplify this
```suggestion
            with open(path, 'rb') as f:
                # NOTE: open the file in bytes mode for safety
                # (prevents decode errors surfacing here)
                contents = f.read()
            contents = contents.replace(
                b'DATAPOINT_API_KEY',
                api_key.encode()
            )
            with open(path, 'wb') as f:
                f.write(contents)
```

Reply:
More efficient reading/writing one line at a time.